Todays Agenda:
1) CNN feature extraction + MLP for classification
2) Transformer Enocder feature classification + MLP for classification
-------------------------------------------------------------------------
3) AleXNet finetune for classification (freeze all layers except last)
4) [ACTIVITY]: Please criticise how can you make the architectures (all 3) more robust


Dataset Structure and Size
Total Images: 60,000 color images.
Split: 50,000 training images and 10,000 testing images.
Classes: 10 distinct, mutually exclusive classes (airplane, automobile, bird, cat, deer, dog, frog, horse, ship, truck).
Balance: The dataset is perfectly balanced, with 6,000 images per class (5,000 for training, 1,000 for testing).
Image Dimensions: Small, low-resolution color images of 32x32 pixels in RGB format

In [1]:
# CNN FEATURE EXTRACTION AND CLASSIFICATION

In [2]:
# Image → CNN → feature vector → MLP → class
# Dataset: CIFAR10

In [1]:
!pip install torch torchvision matplotlib

In [2]:
import torch
import torch.nn as nn
from torchvision import datasets, transforms
from torch.utils.data import DataLoader, random_split

In [3]:


if torch.cuda.is_available():
    device = torch.device("cuda")
elif torch.backends.mps.is_available():
    device = torch.device("mps")
else:
    device = torch.device("cpu")

print("Using device:", device)

Using device: mps


In [4]:
# -----------------------------
# 1. Load Dataset
# -----------------------------

# Convert images to tensor
transform = transforms.ToTensor()

# Training dataset (50k images)
train_dataset = datasets.CIFAR10(
    root="./data",
    train=True,
    download=True,
    transform=transform
)

# Validation dataset (CIFAR test set)
val_dataset = datasets.CIFAR10(
    root="./data",
    train=False,
    download=True,
    transform=transform
)

# Data loaders
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True, pin_memory=False)
val_loader = DataLoader(val_dataset, batch_size=64, pin_memory=False)

/Users/mruksad/Documents/IIT-B/epgd/thirdSem/AI-Ml_in_practice/local/lib/python3.14/site-packages/torchvision/datasets/cifar.py:83: VisibleDeprecationWarning: dtype(): align should be passed as Python or NumPy boolean but got `align=0`. Did you mean to pass a tuple to create a subarray type? (Deprecated NumPy 2.4)
  entry = pickle.load(f, encoding="latin1")


In [5]:
# Define a CNN feature extractor class that inherits from PyTorch's nn.Module
class CNNFeatureExtractor(nn.Module):

    # Constructor function where layers are defined
    def __init__(self):

        # Call the constructor of the parent class (nn.Module)
        super().__init__()

        # Define a sequence of layers that will be applied one after another
        self.cnn = nn.Sequential(

            # First convolution layer
            # Input channels = 3 (RGB image)
            # Output channels = 32 (number of filters)
            # kernel_size = 3 means a 3x3 filter
            # padding = 1 keeps spatial size same (32x32 remains 32x32)
            nn.Conv2d(3, 32, kernel_size=3, padding=1),

            # ReLU activation function
            # Applies max(0,x) to introduce non-linearity
            nn.ReLU(),

            # Max pooling layer
            # kernel_size = 2 downsamples spatial dimension
            # 32x32 becomes 16x16 but 32 channels remain
            nn.MaxPool2d(2),

            # Second convolution layer
            # Takes 32 feature maps as input
            # Produces 64 feature maps
            # kernel_size = 3 looks at local 3x3 spatial regions
            # padding = 1 again keeps spatial dimension same before pooling
            nn.Conv2d(32, 64, kernel_size=3, padding=1),

            # ReLU activation again
            nn.ReLU(),

            # Second pooling layer
            # Reduces spatial dimension again
            # 16x16 becomes 8x8 but 64 channles remain
            nn.MaxPool2d(2)
        ) # shapre of output tensor after this block is [Batch size, 64 channels,8 height, 8 width]


    # Forward pass defines how data flows through the network
    def forward(self, x):

        # Pass input image through the CNN layers defined above
        x = self.cnn(x)

        # At this point tensor shape is:
        # [Batch size, 64 channels, 8 height, 8 width]

        # Flatten the feature maps into a vector
        # x.size(0) is actually the batch size, from the prvious shape [Batch size, 64, 8, 8], 
        # we want to reshape it to [Batch size, 4096]
        # -1 automatically computes remaining dimension (64*8*8 = 4096)
        x = x.view(x.size(0), -1)
        """ view is a fast reshape that (when possible) returns a view on the same underlying memory.

    One important caveat with view is that it requires the tensor to be contiguous in memory.

    If you ever do operations like transpose/permute before flattening, view can error.
    Safer alternatives you’ll often see:

    x = torch.flatten(x, 1) (flatten all dims except batch)
    x = x.reshape(x.size(0), -1) (works even if non-contiguous; may copy if needed) """

        # Return the flattened feature vector
        # Shape becomes [Batch size, 4096]
        return x

In [6]:
# -----------------------------
# 3. MLP Classifier
# -----------------------------

class MLPClassifier(nn.Module):

    def __init__(self):
        super().__init__()

        self.mlp = nn.Sequential(
            nn.Linear(64 * 8 * 8, 256),
            nn.ReLU(),
            nn.Linear(256, 10)
        )

    def forward(self, x):

        return self.mlp(x)

In [7]:
# -----------------------------
# 4. Full Model
# -----------------------------

class Model(nn.Module):

    def __init__(self):
        super().__init__()

        self.encoder = CNNFeatureExtractor()
        self.classifier = MLPClassifier()

    def forward(self, x):

        features = self.encoder(x)

        output = self.classifier(features)

        return output


model = Model().to(device)


In [8]:
# -----------------------------
# 5. Loss and Optimizer
# -----------------------------

criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

In [ ]:
from tqdm import tqdm
from sklearn.metrics import f1_score

# number of epochs
epochs = 50

# store validation losses
val_losses = []

# store validation f1 scores
val_f1_scores = []

for epoch in range(epochs):

    # -----------------------------
    # Training Phase
    # -----------------------------

    # set model to training mode
    model.train()

    # progress bar for training batches
    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs} Training")

    for images, labels in train_bar:

        # forward pass
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)

        # compute training loss
        loss = criterion(outputs, labels)

        # clear previous gradients
        optimizer.zero_grad()

        # compute gradients
        loss.backward()

        # update model weights
        optimizer.step()

        # display current loss in tqdm bar
        train_bar.set_postfix(train_loss=loss.item())


    # -----------------------------
    # Validation Phase
    # -----------------------------

    # set model to evaluation mode
    model.eval()

    val_loss = 0

    # store predictions and labels
    all_preds = []
    all_labels = []

    # disable gradient computation
    with torch.no_grad():

        # validation progress bar
        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{epochs} Validation")

        for images, labels in val_bar:

            # forward pass
            images, labels = images.to(device), labels.to(device)
            outputs = model(images)

            # compute loss
            loss = criterion(outputs, labels)

            val_loss += loss.item()

            # get predicted class
            preds = torch.argmax(outputs, dim=1)

            # store predictions and labels for F1 score
            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())


    # average validation loss
    val_loss = val_loss / len(val_loader)

    # compute F1 score
    f1 = f1_score(all_labels, all_preds, average="macro")

    # store metrics
    val_losses.append(val_loss)
    val_f1_scores.append(f1)

    print(f"\nEpoch {epoch+1}")
    print(f"Validation Loss: {val_loss:.4f}")
    print(f"Validation F1 Score: {f1:.4f}")

0.00s - Debugger warning: It seems that frozen modules are being used, which may
0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off
0.00s - to python to disable frozen modules.
0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.
Epoch 1/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 275.24it/s]



Epoch 1
Validation Loss: 1.1637
Validation F1 Score: 0.5766


Epoch 2/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 321.63it/s]



Epoch 2
Validation Loss: 0.9838
Validation F1 Score: 0.6520


Epoch 3/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 336.52it/s]



Epoch 3
Validation Loss: 0.9108
Validation F1 Score: 0.6810


Epoch 4/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 325.58it/s]



Epoch 4
Validation Loss: 0.8926
Validation F1 Score: 0.6810


Epoch 5/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 338.01it/s]



Epoch 5
Validation Loss: 0.9169
Validation F1 Score: 0.6854


Epoch 6/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 328.39it/s]



Epoch 6
Validation Loss: 0.8613
Validation F1 Score: 0.7106


Epoch 7/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 330.06it/s]



Epoch 7
Validation Loss: 0.8827
Validation F1 Score: 0.7075


Epoch 8/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 334.86it/s]



Epoch 8
Validation Loss: 0.9256
Validation F1 Score: 0.7099


Epoch 9/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 341.51it/s]



Epoch 9
Validation Loss: 1.0028
Validation F1 Score: 0.7004


Epoch 10/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 342.82it/s]



Epoch 10
Validation Loss: 1.0340
Validation F1 Score: 0.7102


Epoch 11/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 333.16it/s]



Epoch 11
Validation Loss: 1.1104
Validation F1 Score: 0.7118


Epoch 12/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 343.01it/s]



Epoch 12
Validation Loss: 1.2421
Validation F1 Score: 0.7088


Epoch 13/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 352.10it/s]



Epoch 13
Validation Loss: 1.3233
Validation F1 Score: 0.7070


Epoch 14/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 354.20it/s]



Epoch 14
Validation Loss: 1.4186
Validation F1 Score: 0.7107


Epoch 15/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 345.91it/s]



Epoch 15
Validation Loss: 1.5709
Validation F1 Score: 0.6948


Epoch 16/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 329.87it/s]



Epoch 16
Validation Loss: 1.6992
Validation F1 Score: 0.7035


Epoch 17/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 344.34it/s]



Epoch 17
Validation Loss: 1.7757
Validation F1 Score: 0.6999


Epoch 18/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 325.39it/s]



Epoch 18
Validation Loss: 1.8383
Validation F1 Score: 0.6958


Epoch 19/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 339.34it/s]



Epoch 19
Validation Loss: 1.9386
Validation F1 Score: 0.6943


Epoch 20/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 336.70it/s]



Epoch 20
Validation Loss: 2.0217
Validation F1 Score: 0.7005


Epoch 21/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 317.54it/s]



Epoch 21
Validation Loss: 1.9953
Validation F1 Score: 0.6961


Epoch 22/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 345.95it/s]



Epoch 22
Validation Loss: 2.1579
Validation F1 Score: 0.6955


Epoch 23/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 352.23it/s]



Epoch 23
Validation Loss: 2.2815
Validation F1 Score: 0.6982


Epoch 24/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 340.90it/s]



Epoch 24
Validation Loss: 2.2567
Validation F1 Score: 0.6943


Epoch 25/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 346.09it/s]



Epoch 25
Validation Loss: 2.3257
Validation F1 Score: 0.6975


Epoch 26/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 352.09it/s]



Epoch 26
Validation Loss: 2.3178
Validation F1 Score: 0.6986


Epoch 27/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 344.15it/s]



Epoch 27
Validation Loss: 2.4680
Validation F1 Score: 0.6964


Epoch 28/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 335.58it/s]



Epoch 28
Validation Loss: 2.4710
Validation F1 Score: 0.6993


Epoch 29/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 348.73it/s]



Epoch 29
Validation Loss: 2.5529
Validation F1 Score: 0.6939


Epoch 30/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 352.68it/s]



Epoch 30
Validation Loss: 2.5720
Validation F1 Score: 0.6974


Epoch 31/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 343.45it/s]



Epoch 31
Validation Loss: 2.5718
Validation F1 Score: 0.6926


Epoch 32/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 345.53it/s]



Epoch 32
Validation Loss: 2.7528
Validation F1 Score: 0.6931


Epoch 33/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 343.73it/s]



Epoch 33
Validation Loss: 2.7940
Validation F1 Score: 0.6905


Epoch 34/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 340.83it/s]



Epoch 34
Validation Loss: 2.7286
Validation F1 Score: 0.6927


Epoch 35/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 341.36it/s]



Epoch 35
Validation Loss: 2.7964
Validation F1 Score: 0.6909


Epoch 36/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 346.36it/s]



Epoch 36
Validation Loss: 2.9460
Validation F1 Score: 0.6969


Epoch 37/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 339.17it/s]



Epoch 37
Validation Loss: 2.9400
Validation F1 Score: 0.6910


Epoch 38/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 346.25it/s]



Epoch 38
Validation Loss: 2.9421
Validation F1 Score: 0.6942


Epoch 39/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 343.02it/s]



Epoch 39
Validation Loss: 2.8993
Validation F1 Score: 0.6941


Epoch 40/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 344.70it/s]



Epoch 40
Validation Loss: 2.9592
Validation F1 Score: 0.6964


Epoch 41/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 344.70it/s]



Epoch 41
Validation Loss: 3.0810
Validation F1 Score: 0.6883


Epoch 42/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 336.52it/s]



Epoch 42
Validation Loss: 3.0352
Validation F1 Score: 0.6944


Epoch 43/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 337.51it/s]



Epoch 43
Validation Loss: 3.2992
Validation F1 Score: 0.6846


Epoch 44/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 347.76it/s]



Epoch 44
Validation Loss: 3.1332
Validation F1 Score: 0.6914


Epoch 45/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 338.70it/s]



Epoch 45
Validation Loss: 3.1780
Validation F1 Score: 0.6960


Epoch 46/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 342.20it/s]



Epoch 46
Validation Loss: 3.1940
Validation F1 Score: 0.6927


Epoch 47/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 318.63it/s]



Epoch 47
Validation Loss: 3.2622
Validation F1 Score: 0.6876


Epoch 48/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 339.23it/s]



Epoch 48
Validation Loss: 3.3316
Validation F1 Score: 0.6877


Epoch 49/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 310.25it/s]



Epoch 49
Validation Loss: 3.3717
Validation F1 Score: 0.6937


Epoch 50/50 Validation: 100%|██████████| 157/157 [00:00<00:00, 341.82it/s]


Epoch 50
Validation Loss: 3.3737
Validation F1 Score: 0.6931


Your final line means: after training 50 epochs, on the validation (CIFAR‑10 test) set your model gets **macro‑F1 ≈ 0.693** and **cross‑entropy loss ≈ 3.37**.

**Why do Validation Loss and F1 move differently (and fluctuate)?**
- **They measure different things.**
  - **F1 (macro)** depends only on the **predicted class** (argmax). If the predicted label stays the same, F1 barely changes.
  - **Cross‑entropy loss** depends on the **predicted probabilities**. If the model becomes **overconfident** (very high probability on its chosen class), then:
    - Correct predictions don’t reduce loss much further,
    - A smaller number of **wrong-but-confident** predictions can blow up the average loss.
  - That’s why you can see **F1 stay ~flat (~0.69)** while **val loss grows big (~3.37)**: the model is getting *more confident*, not necessarily *more correct*.
- **Fluctuations are normal** with minibatch SGD/Adam: each epoch lands in a slightly different spot, so metrics wiggle. (Also: the `train_loss=...` shown in the tqdm bar is usually just the *last batch’s loss*, which is noisy and not comparable across epochs.)

**What your run likely indicates**
- You learned useful features early (your earlier epochs had much lower val loss and higher F1).
- Continuing to epoch 50 pushed the model into **overfitting / poor calibration**: it memorizes training data and becomes too confident, so validation loss increases even if F1 doesn’t collapse.

**Best practices to handle this**
- **Early stopping + “best checkpoint”**
  - Stop when validation loss stops improving (patience like 5–10 epochs).
  - Save the model with **best validation loss** (stable) or **best validation F1** (task metric). Don’t just keep the last epoch.
- **Learning-rate scheduling**
  - Reduce LR when validation loss plateaus (e.g., `ReduceLROnPlateau`) or use cosine decay. This often prevents the late-epoch “confidence explosion.”
- **Regularization to reduce overconfidence**
  - Add **weight decay** (L2) to Adam/SGD.
  - Use **label smoothing** in cross entropy.
  - Add **dropout** in the MLP (and/or BatchNorm in CNN) if you extend the architecture.
- **Better signal/plots**
  - Track and plot **average train loss per epoch** (not last batch), plus val loss and val F1. The curves will make the overfitting point obvious.
- **Data augmentation (big win on CIFAR‑10)**
  - Random crop + horizontal flip typically improves generalization and reduces overfitting.


In [12]:
#Transformer Enocder feature classification + MLP for classification

In [13]:
# Transformer Enocder feature classification + MLP for classification

In [ ]:
class TransformerFeatureExtractor(nn.Module):

    def __init__(self):

        super().__init__()

        # Patch embedding
        # Converts image into patch tokens
        # 32x32 image with patch size 4 → 8x8 patches
        self.patch_embed = nn.Conv2d(
            in_channels=3,
            out_channels=128,
            kernel_size=4,
            stride=4
        ) 

        # Transformer encoder layer
        encoder_layer = nn.TransformerEncoderLayer(
            d_model=128,       # embedding dimension
            nhead=4,           # number of attention heads
            dim_feedforward=256,
            batch_first=True
        )

        # Stack multiple transformer layers
        self.transformer = nn.TransformerEncoder(
            encoder_layer,
            num_layers=2
        )

    def forward(self, x):

        # x shape: [B,3,32,32]

        x = self.patch_embed(x)

        # shape becomes
        # [B,128,8,8]

        x = x.flatten(2)

        # [B,128,64]

        x = x.transpose(1,2)

        # [B,64,128]
        # sequence length = 64 patches

        x = self.transformer(x)

        # mean pooling across patches
        x = x.mean(dim=1)

        # output shape
        # [B,128]

        return x

In [15]:
class MLPClassifier(nn.Module):

    def __init__(self):

        super().__init__()

        self.mlp = nn.Sequential(
            nn.Linear(128, 256),
            nn.ReLU(),
            nn.Linear(256, 10)
        )

    def forward(self, x):

        return self.mlp(x)

In [ ]:
class Model(nn.Module):

    def __init__(self):

        super().__init__()

        self.encoder = TransformerFeatureExtractor()

        self.classifier = MLPClassifier()

    def forward(self, x):

        features = self.encoder(x)

        output = self.classifier(features)

        return output


model = Model().to

In [17]:
criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

In [ ]:
epochs = 50

val_losses = []
val_f1_scores = []

for epoch in range(epochs):

    # -------------------------
    # Training
    # -------------------------

    model.train()

    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs} Training")

    for images, labels in train_bar:
        images, labels = images.to(device), labels.to(device)
        outputs = model(images)

        loss = criterion(outputs, labels)

        optimizer.zero_grad()

        loss.backward()

        optimizer.step()

        train_bar.set_postfix(train_loss=loss.item())


    # -------------------------
    # Validation
    # -------------------------

    model.eval()

    val_loss = 0

    all_preds = []
    all_labels = []

    with torch.no_grad():

        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{epochs} Validation")

        for images, labels in val_bar:
            images, labels = images.to(device), labels.to(device)
            outputs = model(images)

            loss = criterion(outputs, labels)

            val_loss += loss.item()

            preds = torch.argmax(outputs, dim=1)

            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())


    val_loss = val_loss / len(val_loader)

    f1 = f1_score(all_labels, all_preds, average="macro")

    val_losses.append(val_loss)
    val_f1_scores.append(f1)

    print(f"\nEpoch {epoch+1}")
    print(f"Validation Loss: {val_loss:.4f}")
    print(f"Validation F1 Score: {f1:.4f}")

Epoch 1/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.86it/s]



Epoch 1
Validation Loss: 1.5980
Validation F1 Score: 0.3863


Epoch 2/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.35it/s]



Epoch 2
Validation Loss: 1.4553
Validation F1 Score: 0.4668


Epoch 3/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.94it/s]



Epoch 3
Validation Loss: 1.3689
Validation F1 Score: 0.5066


Epoch 4/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.88it/s]



Epoch 4
Validation Loss: 1.2892
Validation F1 Score: 0.5317


Epoch 5/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 13.07it/s]



Epoch 5
Validation Loss: 1.2912
Validation F1 Score: 0.5254


Epoch 6/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.77it/s]



Epoch 6
Validation Loss: 1.2763
Validation F1 Score: 0.5352


Epoch 7/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.87it/s]



Epoch 7
Validation Loss: 1.2113
Validation F1 Score: 0.5668


Epoch 8/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.22it/s]



Epoch 8
Validation Loss: 1.2296
Validation F1 Score: 0.5522


Epoch 9/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.83it/s]



Epoch 9
Validation Loss: 1.1804
Validation F1 Score: 0.5690


Epoch 10/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.24it/s]



Epoch 10
Validation Loss: 1.1857
Validation F1 Score: 0.5797


Epoch 11/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.78it/s]



Epoch 11
Validation Loss: 1.1525
Validation F1 Score: 0.5774


Epoch 12/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.74it/s]



Epoch 12
Validation Loss: 1.1195
Validation F1 Score: 0.6035


Epoch 13/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.70it/s]



Epoch 13
Validation Loss: 1.1173
Validation F1 Score: 0.6031


Epoch 14/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.37it/s]



Epoch 14
Validation Loss: 1.0929
Validation F1 Score: 0.6166


Epoch 15/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 13.06it/s]



Epoch 15
Validation Loss: 1.0793
Validation F1 Score: 0.6182


Epoch 16/50 Validation: 100%|██████████| 157/157 [00:13<00:00, 12.05it/s]



Epoch 16
Validation Loss: 1.1039
Validation F1 Score: 0.6152


Epoch 17/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.66it/s]



Epoch 17
Validation Loss: 1.1154
Validation F1 Score: 0.6078


Epoch 18/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.63it/s]



Epoch 18
Validation Loss: 1.0775
Validation F1 Score: 0.6275


Epoch 19/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.81it/s]



Epoch 19
Validation Loss: 1.0670
Validation F1 Score: 0.6268


Epoch 20/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.87it/s]



Epoch 20
Validation Loss: 1.0855
Validation F1 Score: 0.6237


Epoch 21/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.78it/s]



Epoch 21
Validation Loss: 1.1020
Validation F1 Score: 0.6174


Epoch 22/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.48it/s]



Epoch 22
Validation Loss: 1.0363
Validation F1 Score: 0.6424


Epoch 23/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 13.01it/s]



Epoch 23
Validation Loss: 1.0974
Validation F1 Score: 0.6280


Epoch 24/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.91it/s]



Epoch 24
Validation Loss: 1.0589
Validation F1 Score: 0.6403


Epoch 25/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 13.00it/s]



Epoch 25
Validation Loss: 1.0533
Validation F1 Score: 0.6378


Epoch 26/50 Validation: 100%|██████████| 157/157 [00:11<00:00, 13.24it/s]



Epoch 26
Validation Loss: 1.0660
Validation F1 Score: 0.6400


Epoch 27/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.77it/s]



Epoch 27
Validation Loss: 1.0880
Validation F1 Score: 0.6406


Epoch 28/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.63it/s]



Epoch 28
Validation Loss: 1.1146
Validation F1 Score: 0.6259


Epoch 29/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.90it/s]



Epoch 29
Validation Loss: 1.1199
Validation F1 Score: 0.6363


Epoch 30/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.84it/s]



Epoch 30
Validation Loss: 1.0444
Validation F1 Score: 0.6478


Epoch 31/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 13.01it/s]



Epoch 31
Validation Loss: 1.1269
Validation F1 Score: 0.6331


Epoch 32/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.85it/s]



Epoch 32
Validation Loss: 1.0755
Validation F1 Score: 0.6426


Epoch 33/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.59it/s]



Epoch 33
Validation Loss: 1.1179
Validation F1 Score: 0.6262


Epoch 34/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.77it/s]



Epoch 34
Validation Loss: 1.0960
Validation F1 Score: 0.6371


Epoch 35/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.92it/s]



Epoch 35
Validation Loss: 1.1115
Validation F1 Score: 0.6378


Epoch 36/50 Validation: 100%|██████████| 157/157 [00:12<00:00, 12.72it/s]



Epoch 36
Validation Loss: 1.0614
Validation F1 Score: 0.6538


Epoch 37/50 Validation: 100%|██████████| 157/157 [00:14<00:00, 10.88it/s]



Epoch 37
Validation Loss: 1.1322
Validation F1 Score: 0.6367


Epoch 38/50 Training:  12%|█▏        | 91/782 [00:34<04:25,  2.60it/s, train_loss=0.616]


KeyboardInterrupt: 

In [ ]:
# LETS WORK ON ALEXNET FREEZE ALL LAYER AND UNFREEZE THE LAST LAYER, AS USUAL BE ACTIVE IN CRITICISING THE CODE

In [ ]:
import torch
import torch.nn as nn
from torchvision import datasets, transforms, models
from torch.utils.data import DataLoader
from tqdm import tqdm
from sklearn.metrics import f1_score

In [ ]:
# transform images
transform = transforms.Compose([
    transforms.Resize((224,224)),   # resize CIFAR image
    transforms.ToTensor()
])

# training dataset
train_dataset = datasets.CIFAR10(
    root="./data",
    train=True,
    download=True,
    transform=transform
)

# validation dataset (CIFAR test set)
val_dataset = datasets.CIFAR10(
    root="./data",
    train=False,
    download=True,
    transform=transform
)

# dataloaders
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True, pin_memory=True)
val_loader = DataLoader(val_dataset, batch_size=64, pin_memory=True)

In [ ]:
# load pretrained alexnet
model = models.alexnet(pretrained=True)

In [ ]:
for param in model.parameters():
    param.requires_grad = False

In [ ]:
model.classifier[6] = nn.Linear(4096, 10)
#ONLY THIS LAYER WILL TRAIN 4096 → 10

In [ ]:
criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.Adam(model.classifier[6].parameters(), lr=0.001)

In [ ]:
# select GPU if available, otherwise CPU
device = torch.device(device)

print("Using device:", device)

Using device: cuda


In [ ]:
model = model.to(device)

In [ ]:
epochs = 10

val_losses = []
val_f1_scores = []

for epoch in range(epochs):

    # -------------------------
    # Training
    # -------------------------

    model.train()

    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs} Training")

    for images, labels in train_bar:

        # move data to GPU
        images = images.to(device)
        labels = labels.to(device)

        outputs = model(images)

        loss = criterion(outputs, labels)

        optimizer.zero_grad()

        loss.backward()

        optimizer.step()

        train_bar.set_postfix(train_loss=loss.item())


    # -------------------------
    # Validation
    # -------------------------

    model.eval()

    val_loss = 0

    all_preds = []
    all_labels = []

    with torch.no_grad():

        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{epochs} Validation")

        for images, labels in val_bar:

            # move validation data to GPU
            images = images.to(device)
            labels = labels.to(device)

            outputs = model(images)

            loss = criterion(outputs, labels)

            val_loss += loss.item()

            preds = torch.argmax(outputs, dim=1)

            # move predictions back to CPU for sklearn
            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())


    val_loss = val_loss / len(val_loader)

    f1 = f1_score(all_labels, all_preds, average="macro")

    val_losses.append(val_loss)
    val_f1_scores.append(f1)

    print(f"\nEpoch {epoch+1}")
    print(f"Validation Loss: {val_loss:.4f}")
    print(f"Validation F1 Score: {f1:.4f}")

Epoch 1/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.04it/s]



Epoch 1
Validation Loss: 0.9411
Validation F1 Score: 0.6715


Epoch 2/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.11it/s]



Epoch 2
Validation Loss: 0.8987
Validation F1 Score: 0.6811


Epoch 3/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.19it/s]



Epoch 3
Validation Loss: 0.8902
Validation F1 Score: 0.6807


Epoch 4/10 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.44it/s]



Epoch 4
Validation Loss: 0.8731
Validation F1 Score: 0.6904


Epoch 5/10 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.42it/s]



Epoch 5
Validation Loss: 0.8945
Validation F1 Score: 0.6766


Epoch 6/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.09it/s]



Epoch 6
Validation Loss: 0.8743
Validation F1 Score: 0.6878


Epoch 7/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.19it/s]



Epoch 7
Validation Loss: 0.9184
Validation F1 Score: 0.6650


Epoch 8/10 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.36it/s]



Epoch 8
Validation Loss: 0.8813
Validation F1 Score: 0.6837


Epoch 9/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.26it/s]



Epoch 9
Validation Loss: 0.8546
Validation F1 Score: 0.6949


Epoch 10/10 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.20it/s]


Epoch 10
Validation Loss: 0.8418
Validation F1 Score: 0.7050


In [ ]:
# LETS WORK ON ALEXNET UNFROZEN LAYERS, AS USUAL BE ACTIVE IN CRITICISING THE CODE

In [ ]:
import torch
import torch.nn as nn
from torchvision import datasets, transforms, models
from torch.utils.data import DataLoader
from tqdm import tqdm
from sklearn.metrics import f1_score

In [ ]:
# transform images
transform = transforms.Compose([
    transforms.Resize((224,224)),   # resize CIFAR image
    transforms.ToTensor()
])

# training dataset
train_dataset = datasets.CIFAR10(
    root="./data",
    train=True,
    download=True,
    transform=transform
)

# validation dataset (CIFAR test set)
val_dataset = datasets.CIFAR10(
    root="./data",
    train=False,
    download=True,
    transform=transform
)

# dataloaders
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True, pin_memory=True)
val_loader = DataLoader(val_dataset, batch_size=64, pin_memory=True)

100%|██████████| 170M/170M [00:13<00:00, 12.9MB/s]


In [ ]:
# load pretrained alexnet
model = models.alexnet(pretrained=True)

/usr/local/lib/python3.12/dist-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=AlexNet_Weights.IMAGENET1K_V1`. You can also use `weights=AlexNet_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


Downloading: "https://download.pytorch.org/models/alexnet-owt-7be5be79.pth" to /root/.cache/torch/hub/checkpoints/alexnet-owt-7be5be79.pth


100%|██████████| 233M/233M [00:01<00:00, 180MB/s]


In [ ]:
for param in model.parameters():
    param.requires_grad = True

In [ ]:
criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

In [ ]:
# select GPU if available, otherwise CPU
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

print("Using device:", device)

Using device: cuda


In [ ]:
model = model.to(device)

In [ ]:
epochs = 50

val_losses = []
val_f1_scores = []

for epoch in range(epochs):

    # -------------------------
    # Training
    # -------------------------

    model.train()

    train_bar = tqdm(train_loader, desc=f"Epoch {epoch+1}/{epochs} Training")

    for images, labels in train_bar:

        # move data to GPU
        images = images.to(device)
        labels = labels.to(device)

        outputs = model(images)

        loss = criterion(outputs, labels)

        optimizer.zero_grad()

        loss.backward()

        optimizer.step()

        train_bar.set_postfix(train_loss=loss.item())


    # -------------------------
    # Validation
    # -------------------------

    model.eval()

    val_loss = 0

    all_preds = []
    all_labels = []

    with torch.no_grad():

        val_bar = tqdm(val_loader, desc=f"Epoch {epoch+1}/{epochs} Validation")

        for images, labels in val_bar:

            # move validation data to GPU
            images = images.to(device)
            labels = labels.to(device)

            outputs = model(images)

            loss = criterion(outputs, labels)

            val_loss += loss.item()

            preds = torch.argmax(outputs, dim=1)

            # move predictions back to CPU for sklearn
            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())


    val_loss = val_loss / len(val_loader)

    f1 = f1_score(all_labels, all_preds, average="macro")

    val_losses.append(val_loss)
    val_f1_scores.append(f1)

    print(f"\nEpoch {epoch+1}")
    print(f"Validation Loss: {val_loss:.4f}")
    print(f"Validation F1 Score: {f1:.4f}")

Epoch 1/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.36it/s]



Epoch 1
Validation Loss: 1.3530
Validation F1 Score: 0.4903


Epoch 2/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.48it/s]



Epoch 2
Validation Loss: 0.9335
Validation F1 Score: 0.6667


Epoch 3/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.49it/s]



Epoch 3
Validation Loss: 0.8348
Validation F1 Score: 0.6892


Epoch 4/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.51it/s]



Epoch 4
Validation Loss: 0.7680
Validation F1 Score: 0.7386


Epoch 5/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.49it/s]



Epoch 5
Validation Loss: 0.7586
Validation F1 Score: 0.7415


Epoch 6/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.29it/s]



Epoch 6
Validation Loss: 0.6957
Validation F1 Score: 0.7629


Epoch 7/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.38it/s]



Epoch 7
Validation Loss: 0.7218
Validation F1 Score: 0.7595


Epoch 8/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.40it/s]



Epoch 8
Validation Loss: 0.7057
Validation F1 Score: 0.7702


Epoch 9/50 Validation: 100%|██████████| 157/157 [00:18<00:00,  8.30it/s]



Epoch 9
Validation Loss: 0.6264
Validation F1 Score: 0.7913


Epoch 10/50 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.09it/s]



Epoch 10
Validation Loss: 0.7052
Validation F1 Score: 0.7624


Epoch 11/50 Validation: 100%|██████████| 157/157 [00:19<00:00,  8.16it/s]



Epoch 11
Validation Loss: 0.6656
Validation F1 Score: 0.7822


Epoch 12/50 Validation:  29%|██▉       | 46/157 [00:05<00:13,  7.95it/s]


KeyboardInterrupt: 